### DISCUSSION TITLE
Input looses  focus on first touch

### DISCUSSION THREAD

**Discussion by yannkost** [react-hook-form/react-hook-form#7746] [Q&A] [answered]
I have a problem using react-hook-form with material ui. (I replaced the Textfield in the gif with a normal input, just to check if the behaviour changes. ) I can see that the Controller renders the field again in the console. 

Here the gif where the input looses immediatly focus after clicking on the input. 

![chrome-capture](https://user-images.githubusercontent.com/25202769/152435422-9a0830b4-25e0-4f5a-b561-263bf0f450d3.gif)

And here the code rendering the field & console output. I have searched for quite some time (head banging for 2 evenings now), but I don't understand why, if a field gets the "isTouched" state should be rerendered.  Also I noticed that when focusing, onblur is called immediatly after.

```
 return <Controller
        name={myfield.name}
        control={control}
        rules={myfield.validation}
        render={({
                     field,
                     fieldState,
                     formState
                 }) => {
            console.log("render CONTROLLER", 'field', field, 'touchedFields', formState.touchedFields)
            return (
                <>
                    <Typography
                        sx={{
                            fontWeight: 500,
                            marginLeft: 0.5,
                            marginBottom: 0.5
                        }}
                    >{myfield.displayName}
                        {myfield.validation.required.value ? (<span
                            className={classes.asterix}>*</span>) : null}</Typography>

                    <input type="text" {...field} />
                    {/*<TextField*/}
                    {/*    size={'small'}*/}
                    {/*    {...field}*/}
                    {/*    onFocus={(e) => {*/}
                    {/*        console.log('onFocus',  formState.touchedFields, formState.dirtyFields)*/}
                    {/*    }}*/}
                    {/*    onBlur={(e) => {*/}
                    {/*        console.log('onBlur',  formState.touchedFields, formState.dirtyFields)*/}
                    {/*    }}*/}

                    {/*    value={field.value ? field.value : ''}*/}
                    {/*    fullWidth={myfield.fullWidth}*/}
                    {/*    disabled={myfield.disabled}*/}
                    {/*    className={myfield.disabled ? classes.disabledInput : null}*/}
                    {/*    error={!!fieldState.error}*/}
                    {/*    helperText={fieldState.error ? fieldState.error.message : ''}/>*/}
                </>
            )
        }


        }
    />
```

![image](https://user-images.githubusercontent.com/25202769/152436061-67e3b172-3892-4ff7-af71-ce41c5fe1ce8.png)

Any suggestions are welcome to solve this problem.

**Comment by bluebill1049** [comment]
can you reproduce it in a codesandbox?

**Reply by Moshyfawn** [reply]
I suspect it's smth to do with the MUI modal inside a React portal, so, the reproduction would be handy

**Comment by yannkost** [comment]
I have a sandbox here  (needed some time to set it up). Seems to be working incredibly slowly (Material-ui?) 
But it works in the sandbox!! (Tested it yesterday already). But it doesn't really reflect my setup (nextjs).  But the princip is there. The first form (add) works in my personal project and in the sandbox, focus stays still and all fine. The edit form loses focus on the first click (not on value change, or update or anything, just the first click). But as said, it works in the sandbox  (dependcies of material-ui have same version number). 

https://codesandbox.io/s/cool-hugle-ps3wd?file=/src/App.tsx

**Reply by Moshyfawn** [reply]
To the "working incredibly slowly" point, it's due to you trying to `console.log` the `event` `onBlur` and `onFocus`. I can't remember the explanation, nor do I have a link to it _(ofc I had to bury it somewhere lol)_. You might see
> The operation is insecure.

after logging the `event` and a couple of freezes in CSB.
But it doesn't reflect the issue you're having.

Now, you're saying it's a Next.js project you're seeing the issue in. I tried porting your example to [a Next.js codesanbox](https://codesandbox.io/s/green-wood-gu643?file=/src/RHFromWrapper.tsx). Haven't encountered the issue, but I see quite a few warnings like this one:
> Prop `id` did not match. Server: "E8bs1pn-waQ16M_aZV1qd" Client: "8N-J5FbiI4CiXQ8BKSnQP"

I see that you use [the `nanoid` package](https://www.npmjs.com/package/nanoid) to generate your form IDs. I'd suggest to look into providing stable ids instead of generating random once every render. It might not necessarily fix the issue, but it's the only thing React complaining about that I see atm.

**Comment by yannkost** [comment]
But I will gladly share my project if needed. (Just a little big setup, nextjs,prisma etc..)

**Comment by yannkost** [comment]
@Moshyfawn  Thanks for the reply concerning the id, I have adapted the codesandbox.

@bluebill1049 

I have adapted the codesandbox, it does now behave as I stated at the beginning of the question. When clicking on the edit-button on the table, the form pops up and when focusing the first input field the onblur function gets called immediatly. It seems to be related to the datagrid from material-ui? (Still seems strange).  I left the add button, as the form called by this button does not loose focus.

https://codesandbox.io/s/cool-hugle-ps3wd?file=/src/App.tsx

**Comment by Moshyfawn** [comment]
In the past _(like last couple of weeks lol)_ people have reported a similar issue with their input's focus _(not only MUI)_. And the problem was that they were initializing their components inside the functional component's body, which leads to those inputs being re-initialized every re-render, so the input's `ref` is always different between re-renders.
I'd suggest to check your code against similar logic as the behavior is very similar to what has been described in #7660 and #7530

**Reply by yannkost** [reply]
I have to admit that I'm not realy sure to understand this answer. My Form is an own component, I don't construct it inside the function body. By the way I'm putting some output in the console in the corresponding components body and it does not get rerendered between the focus-blur call. 

![image](https://user-images.githubusercontent.com/25202769/152580242-c0bc90cf-874e-467a-97f2-42d4db59115b.png)

As we can see in the image,  for ex. render FieldsGenerator only gets called once. But between the focus / blur function no rerender takes place.

**Comment by yannkost** [comment, marked as answer]
I have found out that material-ui has some issues with the click event triggered from a Texfield in a Dialog when there is a GridActionCellItem.

[Focus gets reset to GridActionCellItem when clicking TextField inside Dialog](https://github.com/mui-org/material-ui-x/issues/3158)

The (temp?) fix is to add `e.stopPropagation()` at the onClick event on the Textfield.

My thanks in the meanwhile @bluebill1049  & @Moshyfawn  for taking some time to look at the problem.

**Reply by Moshyfawn** [reply]
Ah, makes sense! 

Thank you for the insight!

**Reply by bluebill1049** [reply]
Thanks goes to @Moshyfawn 🙏

**Comment by stephent** [comment]
On the off chance anyone else ends up here while researching why their form fields no longer retain focus, in my case it appears to be caused by some change in `react-tabs` 4.x (my RHF form is contained within a `react-tabs` `<TabPanel>`). Downgrading to 3.2.3 resolved the mystery focus issues - not sure of root cause as yet.
